src/v/cloud_storage/inventory/inv_consumer.cc

```diff
@@ -0,0 +1,250 @@
+        vlog(cst_log.error, "Failed to consume inventory report: {}", ep);
+        co_await stop();
+        std::rethrow_exception(ep);
+    }
+}
+
+ss::future<> inventory_consumer::do_consume(
+  ss::input_stream<char> s, gzip_compressed_t is_report_compressed) {
+    report_parser parser{std::move(s), is_report_compressed};
+    co_await parser.consume([this](const auto& path) {
+        // The row is expected to be in the form:
+        // "bucket","path"
+        auto pieces = parse_row(path);
+        // If a row is malformed we skip it and do not abort processing. This
+        // way the rest of the entries are processed, and if there was a path in
+        // this row it will be recorded as missing.
+        if (pieces.size() != 2) {
+            vlog(cst_log.warn, "unexpected row in inventory report: {}", path);
+            return ss::make_ready_future();
+        }
+        return process_path(pieces[1]);
+    });
+}
+
+std::vector<ss::sstring> inventory_consumer::parse_row(ss::sstring row) {
+    std::vector<ss::sstring> pieces;
+    for (auto p : row | views::split(',') | views::transform(trim_string)) {
+        pieces.emplace_back(p);
+    }
+    return pieces;
+}
+
+ss::future<> inventory_consumer::process_path(ss::sstring path) {
+    if (auto maybe_ntp = ntp_from_path(path);
+        _ntps.contains(maybe_ntp.value())) {
+        const auto& ntp = maybe_ntp.value();
+        auto hash = xxhash_64(path.data(), path.size());
+
+        if (!_path_hashes.contains(ntp)) {
+            _path_hashes.insert({ntp, fragmented_vector<uint64_t>{}});
+        }
+
+        _path_hashes[ntp].push_back(hash);
+        _total_size += sizeof(hash);
+
+        if (_total_size >= _max_hash_size_in_memory) {
+            co_await flush();
+        }
+    }
+    co_return;
+}
+
+ss::future<>
+inventory_consumer::flush(inventory_consumer::write_all_t write_all_entries) {
+    vlog(
+      cst_log.trace,
+      "Flushing to disk, write_all_entries: {}",
+      write_all_entries);
+
+    _num_flushes += 1;
+    if (write_all_entries) {
+        for (auto& [ntp, hashes] : _path_hashes) {
+            if (!hashes.empty()) {
+                co_await flush_ntp_hashes(ntp, std::move(hashes));
+            }
+        }
+        co_return;
+    }
+
+    fragmented_vector<flush_entry> entries;
+    ranges::transform(
+      std::make_move_iterator(std::begin(_path_hashes)),
+      std::make_move_iterator(std::end(_path_hashes)),
+      std::back_inserter(entries),
+      [](auto e) {
+          return flush_entry{std::move(e.first), std::move(e.second)};
+      });
+
+    ranges::sort(entries, ranges::greater(), [](const auto& e) {
+        return e.hashes.size();
+    });
+
+    const auto target_size = _max_hash_size_in_memory / 3;
+    vlog(
+      cst_log.trace,
+      "Before flush, size in memory {}, target {}",
+      _total_size,
+      target_size);
+
+    size_t idx = 0;
+    while (idx < entries.size() && _total_size >= target_size) {
```

Comment:
just wondering if it would be simpler to flush and clear the data structures when they reached a certain size rather than trying to remove a partial amount? IIUC the goal i just to create the separate ntp files efficiently?